### ISSUE TITLE
Views are updated twice, which is slow

### ISSUE THREAD

**Issue by eschnett** [LLNL/spack#23424]
I am installing a set of packages via an environment. This environment also defines a view. This view is updated twice in a row instead of only once. Since I'm on an HPC system, updating the view is very slow.

### Steps to reproduce the issue

```console
$ spack --env-dir utils/spack/symmetry-llvm/compilers install --fail-fast
```

My `spack.yaml` is:
```yaml
spack:
  specs:
    - 'gcc @11.1.0 +graphite'
    - 'llvm @12.0.0 +flang'
  concretization: together
  view: /home/eschnetter/src/CarpetX/Cactus/view-compilers-llvm
```

### Error Message

The screen output ends with:
```console
==> Installing llvm-12.0.0-4bkkbbfb6emktggrzedtjuli3k3gmpah
==> No binary for llvm-12.0.0-4bkkbbfb6emktggrzedtjuli3k3gmpah found: installing from source
==> Using cached archive: /home/eschnetter/src/CarpetX/spack/var/spack/cache/_source-cache/archive/8e/8e6c99e482bb16a450165176c2d881804976a2d770e0445af4375e78a1fbf19c.tar.gz
==> No patches needed for llvm
==> llvm: Executing phase: 'cmake'
==> llvm: Executing phase: 'build'
==> llvm: Executing phase: 'install'
==> llvm: Successfully installed llvm-12.0.0-4bkkbbfb6emktggrzedtjuli3k3gmpah
  Fetch: 3.74s.  Build: 35m 49.76s.  Total: 35m 53.49s.
[+] /home/eschnetter/src/CarpetX/spack/opt/spack/linux-ubuntu18.04-skylake_avx512/gcc-10.1.0/llvm-12.0.0-4bkkbbfb6emktggrzedtjuli3k3gmpah
==> Updating view at /home/eschnetter/src/CarpetX/Cactus/view-compilers-llvm
==> Updating view at /home/eschnetter/src/CarpetX/Cactus/view-compilers-llvm
```

Note the duplicate message about `Updating view` at the end. The first update takes several minutes, the second about one minute.

I usually run separate concretization step before (`spack --env-dir utils/spack/symmetry/compilers concretize --force`). This also updates the view (which is useless since the packages haven't been built yet), and which is also slow. It seems Spack just likes updating views.

The complete output (for a run when all packages are already built) is:
```
$ spack -d --env-dir utils/spack/symmetry-llvm/compilers install -j40 --fail-fast
==> [2021-05-04-10:19:33.045969] Imported install from built-in commands
==> [2021-05-04-10:19:33.050547] Reading config file /home/eschnetter/src/CarpetX/Cactus/utils/spack/symmetry-llvm/compilers/spack.yaml
==> [2021-05-04-10:19:33.055440] Imported install from built-in commands
==> [2021-05-04-10:19:33.204200] Installing environment /home/eschnetter/src/CarpetX/Cactus/utils/spack/symmetry-llvm/compilers
==> [2021-05-04-10:19:33.204302] Assessing installation status of environment packages
==> [2021-05-04-10:19:33.205976] DATABASE LOCK TIMEOUT: 3s
==> [2021-05-04-10:19:33.206099] PACKAGE LOCK TIMEOUT: No timeout
==> [2021-05-04-10:19:33.306921] Reading config file /home/eschnetter/src/CarpetX/spack/etc/spack/defaults/repos.yaml
==> [2021-05-04-10:19:38.257678] All of the packages are already installed
==> [2021-05-04-10:19:38.257766] Regenerating environment views for /home/eschnetter/src/CarpetX/Cactus/utils/spack/symmetry-llvm/compilers
==> [2021-05-04-10:19:38.409623] TEMPORARY DIRECTORY CREATED [/home/eschnetter/src/CarpetX/Cactus/tmp0pp7eenb]
==> [2021-05-04-10:19:38.435123] DIRECTORY MOVED [src=/home/eschnetter/src/CarpetX/Cactus/view-compilers-llvm, dest=/home/eschnetter/src/CarpetX/Cactus/tmp0pp7eenb]
==> [2021-05-04-10:19:38.435971] Updating view at /home/eschnetter/src/CarpetX/Cactus/view-compilers-llvm
==> [2021-05-04-10:20:01.808145] Copying /home/eschnetter/src/CarpetX/spack/opt/spack/linux-ubuntu18.04-skylake_avx512/gcc-10.1.0/python-3.8.10-nijjgt66pnyn223m4fv5gf6cmoiphgmo/bin/idle3.8 to /home/eschnetter/src/CarpetX/Cactus/view-compilers-llvm/bin/idle3.8
==> [2021-05-04-10:20:01.813870] 'file' '-b' '-h' '/home/eschnetter/src/CarpetX/spack/opt/spack/linux-ubuntu18.04-skylake_avx512/gcc-10.1.0/python-3.8.10-nijjgt66pnyn223m4fv5gf6cmoiphgmo/bin/idle3.8'
==> [2021-05-04-10:20:01.830301] FILTER FILE: /home/eschnetter/src/CarpetX/Cactus/view-compilers-llvm/bin/idle3.8 [replacing "/home/eschnetter/src/CarpetX/spack/opt/spack/linux-ubuntu18.04-skylake_avx512/gcc-10.1.0/python-3.8.10-nijjgt66pnyn223m4fv5gf6cmoiphgmo"]
==> [2021-05-04-10:20:01.848622] Copying /home/eschnetter/src/CarpetX/spack/opt/spack/linux-ubuntu18.04-skylake_avx512/gcc-10.1.0/python-3.8.10-nijjgt66pnyn223m4fv5gf6cmoiphgmo/bin/pydoc3.8 to /home/eschnetter/src/CarpetX/Cactus/view-compilers-llvm/bin/pydoc3.8
==> [2021-05-04-10:20:01.853045] 'file' '-b' '-h' '/home/eschnetter/src/CarpetX/spack/opt/spack/linux-ubuntu18.04-skylake_avx512/gcc-10.1.0/python-3.8.10-nijjgt66pnyn223m4fv5gf6cmoiphgmo/bin/pydoc3.8'
==> [2021-05-04-10:20:01.868170] FILTER FILE: /home/eschnetter/src/CarpetX/Cactus/view-compilers-llvm/bin/pydoc3.8 [replacing "/home/eschnetter/src/CarpetX/spack/opt/spack/linux-ubuntu18.04-skylake_avx512/gcc-10.1.0/python-3.8.10-nijjgt66pnyn223m4fv5gf6cmoiphgmo"]
==> [2021-05-04-10:20:01.881620] Copying /home/eschnetter/src/CarpetX/spack/opt/spack/linux-ubuntu18.04-skylake_avx512/gcc-10.1.0/python-3.8.10-nijjgt66pnyn223m4fv5gf6cmoiphgmo/bin/python3.8 to /home/eschnetter/src/CarpetX/Cactus/view-compilers-llvm/bin/python3.8
==> [2021-05-04-10:20:01.885688] 'file' '-b' '-h' '/home/eschnetter/src/CarpetX/spack/opt/spack/linux-ubuntu18.04-skylake_avx512/gcc-10.1.0/python-3.8.10-nijjgt66pnyn223m4fv5gf6cmoiphgmo/bin/python3.8'
==> [2021-05-04-10:20:01.893335] Copying /home/eschnetter/src/CarpetX/spack/opt/spack/linux-ubuntu18.04-skylake_avx512/gcc-10.1.0/python-3.8.10-nijjgt66pnyn223m4fv5gf6cmoiphgmo/bin/python3.8-gdb.py to /home/eschnetter/src/CarpetX/Cactus/view-compilers-llvm/bin/python3.8-gdb.py
==> [2021-05-04-10:20:01.898470] 'file' '-b' '-h' '/home/eschnetter/src/CarpetX/spack/opt/spack/linux-ubuntu18.04-skylake_avx512/gcc-10.1.0/python-3.8.10-nijjgt66pnyn223m4fv5gf6cmoiphgmo/bin/python3.8-gdb.py'
==> [2021-05-04-10:20:01.915164] FILTER FILE: /home/eschnetter/src/CarpetX/Cactus/view-compilers-llvm/bin/python3.8-gdb.py [replacing "/home/eschnetter/src/CarpetX/spack/opt/spack/linux-ubuntu18.04-skylake_avx512/gcc-10.1.0/python-3.8.10-nijjgt66pnyn223m4fv5gf6cmoiphgmo"]
==> [2021-05-04-10:20:01.937326] Copying /home/eschnetter/src/CarpetX/spack/opt/spack/linux-ubuntu18.04-skylake_avx512/gcc-10.1.0/python-3.8.10-nijjgt66pnyn223m4fv5gf6cmoiphgmo/bin/2to3-3.8 to /home/eschnetter/src/CarpetX/Cactus/view-compilers-llvm/bin/2to3-3.8
==> [2021-05-04-10:20:01.941756] 'file' '-b' '-h' '/home/eschnetter/src/CarpetX/spack/opt/spack/linux-ubuntu18.04-skylake_avx512/gcc-10.1.0/python-3.8.10-nijjgt66pnyn223m4fv5gf6cmoiphgmo/bin/2to3-3.8'
==> [2021-05-04-10:20:01.989616] FILTER FILE: /home/eschnetter/src/CarpetX/Cactus/view-compilers-llvm/bin/2to3-3.8 [replacing "/home/eschnetter/src/CarpetX/spack/opt/spack/linux-ubuntu18.04-skylake_avx512/gcc-10.1.0/python-3.8.10-nijjgt66pnyn223m4fv5gf6cmoiphgmo"]
==> [2021-05-04-10:20:02.002352] Copying /home/eschnetter/src/CarpetX/spack/opt/spack/linux-ubuntu18.04-skylake_avx512/gcc-10.1.0/python-3.8.10-nijjgt66pnyn223m4fv5gf6cmoiphgmo/bin/python3.8-config to /home/eschnetter/src/CarpetX/Cactus/view-compilers-llvm/bin/python3.8-config
==> [2021-05-04-10:20:02.006342] 'file' '-b' '-h' '/home/eschnetter/src/CarpetX/spack/opt/spack/linux-ubuntu18.04-skylake_avx512/gcc-10.1.0/python-3.8.10-nijjgt66pnyn223m4fv5gf6cmoiphgmo/bin/python3.8-config'
==> [2021-05-04-10:20:02.020341] FILTER FILE: /home/eschnetter/src/CarpetX/Cactus/view-compilers-llvm/bin/python3.8-config [replacing "/home/eschnetter/src/CarpetX/spack/opt/spack/linux-ubuntu18.04-skylake_avx512/gcc-10.1.0/python-3.8.10-nijjgt66pnyn223m4fv5gf6cmoiphgmo"]
==> [2021-05-04-10:20:35.380247] 'file' '-b' '-h' '/home/eschnetter/src/CarpetX/spack/opt/spack/linux-ubuntu18.04-skylake_avx512/gcc-10.1.0/py-setuptools-50.3.2-5jt7tf7tetj4fbx624dgtazapanqkdnd/bin/easy_install'
==> [2021-05-04-10:20:35.396438] FILTER FILE: /home/eschnetter/src/CarpetX/Cactus/view-compilers-llvm/bin/easy_install [replacing "/home/eschnetter/src/CarpetX/spack/opt/spack/linux-ubuntu18.04-skylake_avx512/gcc-10.1.0/python-3.8.10-nijjgt66pnyn223m4fv5gf6cmoiphgmo"]
==> [2021-05-04-10:20:35.414172] 'file' '-b' '-h' '/home/eschnetter/src/CarpetX/spack/opt/spack/linux-ubuntu18.04-skylake_avx512/gcc-10.1.0/py-setuptools-50.3.2-5jt7tf7tetj4fbx624dgtazapanqkdnd/bin/easy_install-3.8'
==> [2021-05-04-10:20:35.427504] FILTER FILE: /home/eschnetter/src/CarpetX/Cactus/view-compilers-llvm/bin/easy_install-3.8 [replacing "/home/eschnetter/src/CarpetX/spack/opt/spack/linux-ubuntu18.04-skylake_avx512/gcc-10.1.0/python-3.8.10-nijjgt66pnyn223m4fv5gf6cmoiphgmo"]
==> [2021-05-04-10:21:21.574638] TEMPORARY DIRECTORY DELETED [/home/eschnetter/src/CarpetX/Cactus/tmp0pp7eenb]
``` 

### Information on your system

$ spack debug report
* **Spack:** 0.16.1-2504-4ecfe7f09e
* **Python:** 3.6.9
* **Platform:** linux-ubuntu18.04-skylake_avx512
* **Concretizer:** original

### Additional information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have searched the issues of this repo and believe this is not a duplicate
- [x] I have run the failing commands in debug mode and reported the output

**Comment by haampie** [comment]
Worse even if you do distributed builds, they get updated N+1 times

**Comment by alalazo** [comment]
I think this was fixed in the meanwhile (65556eb53e184e78aaacce2a9e325ea4507bca2c).
